ipfs-8-to-9/migration/migration.go

```diff
@@ -0,0 +1,110 @@
+// package mg8 contains the code to perform 8-9 repository migration in
+// go-ipfs. This performs a switch to raw multihashes for all keys in the
+// go-ipfs datastore (https://github.com/ipfs/go-ipfs/issues/6815).
+package mg8
+
+import (
+	"errors"
+	"fmt"
+
+	migrate "github.com/ipfs/fs-repo-migrations/go-migrate"
+	lock "github.com/ipfs/fs-repo-migrations/ipfs-1-to-2/repolock"
+	"github.com/ipfs/go-datastore/namespace"
+
+	mfsr "github.com/ipfs/fs-repo-migrations/mfsr"
+	log "github.com/ipfs/fs-repo-migrations/stump"
+	ds "github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-ipfs/plugin/loader"
+	fsrepo "github.com/ipfs/go-ipfs/repo/fsrepo"
+)
+
+// Migration implements the migration described above.
+type Migration struct{}
+
+// Versions returns the current version string for this migration.
+func (m Migration) Versions() string {
+	return "8-to-9"
+}
+
+// Reversible returns false. This migration cannot be reverted, as we do not
+// know which raw hashes were actually CIDv1s. However, things should work all
+// the same as they will be treated as CIDv0s in old versions anyways.
+func (m Migration) Reversible() bool {
+	return false
+}
+
+// Apply runs the migration.
+func (m Migration) Apply(opts migrate.Options) error {
+	log.Verbose = opts.Verbose
+	log.Log("applying %s repo migration", m.Versions())
+
+	log.VLog("locking repo at %q", opts.Path)
+	lk, err := lock.Lock2(opts.Path)
+	if err != nil {
+		return err
+	}
+	defer lk.Close()
+
+	repo := mfsr.RepoPath(opts.Path)
+
+	log.VLog("  - verifying version is '8'")
+	if err := repo.CheckVersion("8"); err != nil {
+		return err
+	}
+
+	log.VLog("  - loading repo configurations")
+	plugins, err := loader.NewPluginLoader(opts.Path)
+	if err != nil {
+		return fmt.Errorf("error loading plugins: %s", err)
+	}
+
+	if err := plugins.Initialize(); err != nil {
+		return fmt.Errorf("error initializing plugins: %s", err)
+	}
+
+	if err := plugins.Inject(); err != nil {
+		return fmt.Errorf("error injecting plugins: %s", err)
+	}
+
+	cfg, err := fsrepo.ConfigAt(opts.Path)
+	if err != nil {
+		return err
+	}
+
+	dsc, err := fsrepo.AnyDatastoreConfig(cfg.Datastore.Spec)
+	if err != nil {
+		return err
+	}
+
+	dstore, err := dsc.Create(opts.Path)
+	if err != nil {
+		return err
+	}
+	defer dstore.Close()
+
+	// TODO: assuming the user has not modified this
+	blocks := namespace.Wrap(dstore, ds.NewKey("/blocks"))
```

Comment:
Could actually process the root namespace and migrate everything that ends with a CIDv1. Do we want that? This would, afaik, migrate providers too.

Reply:
Hm. I'd wouldn't. We'll delete old providers separately. I don't want to migrate _everything_ that looks like a CID.

Reply:
Real question: do we need to migrate the filestore? Probably.

Reply:
Yes, we need, I guess there is a namespace that I haven't found out yet.

That said, we do want to migrate providers too, but can scope for their namespace.

Reply:
> That said, we do want to migrate providers too, but can scope for their namespace.

Why? They'll expire in a day.

Reply:
Because it is effortless to migrate another namespace, but if you tell me nothing is going to break in those 24h, I can live with that.

Reply:
If it's easy, we can go ahead and do it anyways. My main concern is that I don't want to sit their migrating those records

If we're migrating them anyways, I'd consider (don't spend too much time at this) checking to see if they're expired and delete them up-front (should be faster than re-writing them).

Reply:
@Stebalien I did not realize, the problem is there is actually potentially a very large number of provider blocks to migrate? Might also treat this as an additional migration later. Let me look up some code, there are some things I'm unfamiliar with here. Let's keep this migration as only doing `blocks`.

Reply:
It might be, I don't actually know.